### PR TITLE
Add safe helper for charging gas based on word count

### PR DESCRIPTION
**Pull request by jakelang**

### CODE REVIEW

File: src/eei.cpp [hunk 1]
```diff
@@ -1056,4 +1044,11 @@ inline int64_t maxCallGas(int64_t gas) {
     }
     return true;
   }
+
+  void EthereumInterface::safeChargeDataCopy(uint32_t length, bool isExternalCode) {
+    static_assert(GasSchedule::copy <= 3, "Gas charge value has changed!");
+    takeGas((isExternalCode) ? GasSchedule::extcode : GasSchedule::verylow);
```
Comment (axic) [inline src/eei.cpp]: I would just have a `baseCost` or `baseGas` field instead.

**Comment by axic** [comment]
Also need to refactor the code in `log`.

**Comment by jakelang** [comment]
@axic log wouldn't make sense to refactor because the helper would only be used once

**Comment by axic** [comment]
#323 was meged, this needs to be rebased.

File: src/eei.cpp [hunk 2]
```diff
@@ -302,13 +302,7 @@ inline int64_t maxCallGas(int64_t gas) {
...
       HERA_DEBUG << "callDataCopy " << hex << resultOffset << " " << dataOffset << " " << length << dec << "\n";
 
-      ensureCondition(ffs(GasSchedule::copy) + (ffsl(length) - 5) <= 64, OutOfGas, "Gas charge overflow");
-      ensureCondition(
-        numeric_limits<uint64_t>::max() - GasSchedule::verylow >= GasSchedule::copy * ((uint64_t(length) + 31) / 32),
-        OutOfGas,
-        "Gas charge overflow"
-      );
-      takeInterfaceGas(GasSchedule::verylow + GasSchedule::copy * ((uint64_t(length) + 31) / 32));
+      EthereumInterface::safeChargeDataCopy(length, GasSchedule::verylow);
```
Comment (axic) [inline src/eei.cpp]: Don't need `EthereumInterface` here.

File: src/eei.h [hunk 3]
```diff
@@ -107,6 +107,8 @@ struct EthereumInterface : ShellExternalInterface {
   /* Checks if 256 bit value is all zeroes */
   static bool isZeroUint256(evmc_uint256be const& value);
 
+  void safeChargeDataCopy(uint32_t length, uint32_t basecost);
```
Comment (axic) [inline src/eei.h]: Please add a comment describing what it does.

File: src/eei.cpp [hunk 4]
```diff
@@ -1056,4 +1044,12 @@ namespace hera {
     }
     return true;
   }
+
+  //Safely charges gas for callDataCopy, returnDataCopy, codeCopy, and externalCodeCopy
```
Comment (axic) [inline src/eei.cpp]: I meant to put the short description into the header and it should only say something along the lines of "safely (handling overflows) charge variable gas cost based on word size"

File: src/eei.cpp [hunk 5]
```diff
@@ -1056,4 +1044,12 @@ namespace hera {
...
+  void EthereumInterface::safeChargeDataCopy(uint32_t length, uint32_t basecost) {
```
Comment (axic) [inline src/eei.cpp]: Please still change `basecost` to `baseCost`.

Comment (axic) [reply src/eei.cpp]: Need to also add a `static_assert` for `baseCost`.

**Pull request merged**